belgif-rest-problem-validator/src/main/java/io/github/belgif/rest/problem/validation/AbstractRequestValidator.java

```diff
@@ -0,0 +1,447 @@
+     *
+     * @param input the input to validate
+     * @param <T> the input type
+     * @return this RequestValidator
+     */
+    public <T> V require(Input<T> input) {
+        if (input != null) {
+            addValidator(new RequiredInputValidator<>(input));
+        }
+        return getThis();
+    }
+
+    /**
+     * Validate that all the given inputs are present when the given target input is present.
+     *
+     * @param target the target input
+     * @param inputs the inputs that are required when the target input is present
+     * @return this RequestValidator
+     */
+    public V requireIfPresent(Input<?> target, Input<?>... inputs) {
+        addValidator(new RequiredIfPresentValidator(target, Arrays.asList(inputs)));
+        return getThis();
+    }
+
+    /**
+     * Conditionally register input validators.
+     *
+     * @param condition the condition
+     * @param requestValidatorConsumer the RequestValidator consumer that will only be called if condition is true
+     * @return this RequestValidator
+     */
+    public V when(boolean condition, Consumer<V> requestValidatorConsumer) {
+        if (condition) {
+            requestValidatorConsumer.accept(getThis());
+        }
+        return getThis();
+    }
+
+    /**
+     * Validate that the given input value is in the given [min, max] range.
+     *
+     * @param input the input to validate
+     * @param min the minimum (inclusive)
+     * @param max the maximum (inclusive)
+     * @param <T> the input type
+     * @return this RequestValidator
+     */
+    public <T extends Comparable<T>> V range(Input<T> input, T min, T max) {
+        if (input != null && input.getValue() != null) {
+            addValidator(new RangeValidator<>(input, min, max));
+        }
+        return getThis();
+    }
+
+    /**
+     * Validate that the given input value is not less than the given minimum.
+     *
+     * @param input the input to validate
+     * @param min the minimum (inclusive)
+     * @param <T> the input type
+     * @return this RequestValidator
+     */
+    public <T extends Comparable<T>> V minimum(Input<T> input, T min) {
+        return range(input, min, null);
+    }
+
+    /**
+     * Validate that the given input value does not exceed the given maximum.
+     *
+     * @param input the input to validate
+     * @param max the maximum (inclusive)
+     * @param <T> the input type
+     * @return this RequestValidator
+     */
+    public <T extends Comparable<T>> V maximum(Input<T> input, T max) {
+        return range(input, null, max);
+    }
+
+    /**
+     * Register a custom {@link InputValidator}.
+     *
+     * @param validator the custom validator
+     * @return this RequestValidator
+     */
+    public V custom(InputValidator validator) {
+        if (validator != null) {
+            addValidator(validator);
+        }
+        return getThis();
+    }
+
+    protected abstract V getThis();
```

Comment:
Don't you do like assertj by giving the `Class<?> selfType` in the constructor and store the self type by doing `myself = (SELF) selfType.cast(this)`. It requires a cast in the abstract constructor but it's easier to use by the class extending it.

Reply:
Something like this then https://github.com/belgif/rest-problem-java/pull/101/commits/a17cf50c6df559daef7ac42331f00eb4bc506d8d?